src/debug-generator/debug-generator.c

```diff
@@ -20,20 +22,95 @@
 #include "unit-file.h"
 #include "unit-name.h"
 
+typedef enum BreakpointType {
+        BREAKPOINT_PRE_UDEV,
+        BREAKPOINT_PRE_SYSROOT_MOUNT,
+        BREAKPOINT_PRE_SWITCH_ROOT,
+        _BREAKPOINT_TYPE_MAX,
+        _BREAKPOINT_TYPE_INVALID = -EINVAL,
+} BreakpointType;
+
+typedef struct BreakpointInfo {
+        BreakpointType type;
+        const char *name;
+        const char *unit;
+        bool valid_in_initrd;
+        bool valid_in_main_system;
+} BreakpointInfo;
+
 static const char *arg_dest = NULL;
 static char *arg_default_unit = NULL;
 static char **arg_mask = NULL;
 static char **arg_wants = NULL;
 static bool arg_debug_shell = false;
 static char *arg_debug_tty = NULL;
 static char *arg_default_debug_tty = NULL;
+static uint32_t arg_breakpoints = 0;
 
 STATIC_DESTRUCTOR_REGISTER(arg_default_unit, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_mask, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_wants, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_debug_tty, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_default_debug_tty, freep);
 
+static const struct BreakpointInfo breakpoint_info_table[_BREAKPOINT_TYPE_MAX] = {
+        { BREAKPOINT_PRE_UDEV,          "pre-udev",        "breakpoint-pre-udev.service",        true,  true  },
+        { BREAKPOINT_PRE_SYSROOT_MOUNT, "pre-mount",       "breakpoint-pre-mount.service",       true,  false },
+        { BREAKPOINT_PRE_SWITCH_ROOT,   "pre-switch-root", "breakpoint-pre-switch-root.service", true,  false },
+};
+
+static BreakpointType parse_breakpoint_from_string_one(const char *s) {
+        assert(s);
+
+        for (size_t i = 0; i < ELEMENTSOF(breakpoint_info_table); i++)
+                if (streq(breakpoint_info_table[i].name, s))
+                        return breakpoint_info_table[i].type;
+
+        return _BREAKPOINT_TYPE_INVALID;
+}
+
+static int parse_breakpoint_from_string(const char *s, uint32_t *ret_breakpoints) {
+        uint32_t breakpoints = 0;
+
+        assert(ret_breakpoints);
+
+        /* Empty value? set default breakpoint */
+        if (!s) {
+                if (in_initrd())
+                        breakpoints |= (1 << BREAKPOINT_PRE_SWITCH_ROOT);
```

Comment:
i think it would be a bit nicer to encode which one is the default one via another flag in the BreakpointValidity enum: BREAKPOINT_DEFAULT or so. If the empty string is specified you'd look for the first one with that flag set, and use that if you find one.

Reply:
I find a bit over-engineered having to iterate the table looking for this.. isn't it just a known constant? maybe having a define that points to the default? Anyway, if you want it this way, you'll have it.